src/components/core/GeoChart.tsx

```diff
@@ -74,44 +109,165 @@ export class GeoChartInner extends BaseVisualization<IGeoChartInnerProps, {}> {
+            const attributeHeaderItems = getGeoAttributeHeaderItems(executionResult, geoData);
+            let segmentByData: IPushpinCategoryLegendItem[] = [];
+            let sizeData: number[] = [];
+            if (size && isTwoDimensionsData(data)) {
+                sizeData = data[size.index].map(stringToFloat);
+            }
+            if (segmentBy) {
+                const segmentByAttributeHeaderItems = attributeHeaderItems[segmentBy.index];
+                segmentByData = this.getSegmentBySeries(segmentByAttributeHeaderItems);
+                if (this.state === undefined) {
+                    this.state = {
+                        legendItemsEnabled: [...segmentByData],
+                    };
+                } else if (!isEqual(this.state.legendItemsEnabled, segmentByData)) {
+                    this.setState({
+                        legendItemsEnabled: [...segmentByData],
+                    });
+                }
+            }
+            return {
+                geoData,
+                segmentByData,
+                sizeData,
+            };
+        }
+    }
+
+    private getLegendProps(): IGeoChartLegendRendererProps {
+        const {
             config,
+            locale,
+            legendPosition: position,
+            execution: { executionResponse },
+        } = this.props;
+        let legendProps: IGeoChartLegendRendererProps = {
+            config,
+            locale,
+            position,
+            geoData: {},
+        };
+        if (this.geoChartOptions) {
+            const { size, segmentBy } = this.geoChartOptions.geoData;
+            legendProps = { ...legendProps, geoData: { ...this.geoChartOptions.geoData } };
+            if (size) {
+                const sizeLegendOptions = {
+                    sizeData: this.geoChartOptions.sizeData || [],
+                    sizeFormat: getFormatFromExecutionResponse(size.index, executionResponse),
+                };
+                legendProps = { ...legendProps, ...sizeLegendOptions };
+            }
+            if (segmentBy) {
+                const segmentByLegendOptions = {
+                    segmentByData: this.geoChartOptions.segmentByData || [],
+                    onItemClick: this.onLegendItemClick,
+                };
+                legendProps = { ...legendProps, ...segmentByLegendOptions };
+            }
+        }
+        return legendProps;
+    }
+
+    private getChartProps(): IGeoChartRendererProps {
+        const { config, execution, afterRender } = this.props;
+        const segmentByIndex: number = get(this.geoChartOptions, "geoData.segmentBy.index");
+        let selectedSegmentNames: string[];
+        if (segmentByIndex) {
+            selectedSegmentNames = this.state.legendItemsEnabled
+                .filter((item: IPushpinCategoryLegendItem): boolean => item.isVisible)
+                .map((item: IPushpinCategoryLegendItem): string => item.name);
```

Comment:
It could be
```
    if (item.isVisible) {
                    result.push(item.name);
                } 
                return result;
            }, []);
```
to reduce loop times.

Reply:
fixed